src/snowflake/snowpark/_internal/data_source_utils.py

```diff
@@ -0,0 +1,351 @@
+        elif dbms_type == DBMS_TYPE.ORACLE_DB:
+            query = f"""
+                    SELECT COLUMN_NAME, DATA_TYPE, DATA_PRECISION, DATA_SCALE, NULLABLE
+                    FROM USER_TAB_COLUMNS
+                    WHERE TABLE_NAME = '{table}'
+                    """
+            cursor.execute(query)
+            raw_schema = cursor.fetchall()
+            return oracledb_to_snowpark_type(raw_schema)
+        else:
+            raise NotImplementedError(
+                f"Failed to infer Snowpark DataFrame schema from source '{table}'. "
+                f"Currently supported drivers are 'pyodbc' and 'oracledb', but got: '{driver_info}'. "
+                "To avoid auto inference, you can manually specify the Snowpark DataFrame schema using 'custom_schema' in DataFrameReader.dbapi."
+            )
+
+    except Exception as exc:
+        raise SnowparkDataframeReaderException(
+            f"Failed to infer Snowpark DataFrame schema from table '{table}'. To avoid auto inference, you can manually specify the Snowpark DataFrame schema using 'custom_schema' in DataFrameReader.dbapi."
+        ) from exc
+
+
+def data_source_data_to_pandas_df(
+    data: List[Any], schema: StructType, current_db: str, driver_info: str
+) -> pd.DataFrame:
+    columns = [col.name for col in schema.fields]
+    df = pd.DataFrame.from_records(data, columns=columns)
+
+    # convert timestamp and date to string to work around SNOW-1911989
+    df = df.map(
+        lambda x: x.isoformat()
+        if isinstance(x, (datetime.datetime, datetime.date))
+        else x
+    )
+    # convert binary type to object type to work around SNOW-1912094
+    df = df.map(lambda x: x.hex() if isinstance(x, (bytearray, bytes)) else x)
+    return df
+
+
+def generate_select_query(
+    table: str, schema: StructType, dbms: DBMS_TYPE, driver_info: str
+) -> str:
+    if dbms == DBMS_TYPE.ORACLE_DB:
+        cols = []
+        for field in schema.fields:
+            if (
+                isinstance(field.datatype, TimestampType)
+                and field.datatype.tz == TimestampTimeZone.TZ
+            ):
+                cols.append(
+                    f"""TO_CHAR({field.name}, 'YYYY-MM-DD HH24:MI:SS.FF9 TZHTZM')"""
+                )
+            elif (
+                isinstance(field.datatype, TimestampType)
+                and field.datatype.tz == TimestampTimeZone.LTZ
+            ):
+                cols.append(
+                    f"""TO_CHAR({field.name} AT TIME ZONE SESSIONTIMEZONE, 'YYYY-MM-DD HH24:MI:SS.FF9 TZHTZM')"""
+                )
+            else:
+                cols.append(field.name)
+        return f"""select {" , ".join(cols)} from {table}"""
+    elif dbms == DBMS_TYPE.SQL_SERVER_DB or dbms == DBMS_TYPE.SQLITE_DB:
+        return f"select * from {table}"
+    else:
+        raise NotImplementedError(
+            f"currently supported drivers are pyodbc and oracledb, got: {driver_info}"
+        )
```

Comment:
let's remove this error raising part, this gives more flexility to users to try out different db
```suggestion
    else:
        return f"select * from {table}"
```